File: services/kg_cache/cache.go
package main

import (
	"bytes"
	"encoding/json"
	"log"
	"net/http"
	"sync"
	"time"
)

const TTL = 15 * time.Minute

// CacheEntry holds a JSON-encoded DisruptionEvent and its expiry time.
type CacheEntry struct {
	Value     string // JSON-encoded DisruptionEvent
	ExpiresAt time.Time
}

// Cache is a thread-safe in-memory TTL cache.
type Cache struct {
	mu          sync.RWMutex
	entries     map[string]CacheEntry
	webIntelURL string
}

// NewCache creates a new Cache with the given Web_Intelligence_Service callback URL.
func NewCache(webIntelURL string) *Cache {
	return &Cache{
		entries:     make(map[string]CacheEntry),
		webIntelURL: webIntelURL,
	}
}

// Set stores a value under key with a fresh TTL.
// Key format: "{zone_id}:{event_type}"
func (c *Cache) Set(key, value string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.entries[key] = CacheEntry{
		Value:     value,
		ExpiresAt: time.Now().Add(TTL),
	}
}

// Get retrieves a value by key.
// Returns ("", false) if the key is missing or the entry has expired.
// On expiry: evicts the entry and triggers an async re-scrape.
func (c *Cache) Get(key string) (string, bool) {
	// Fast path: read lock to check existence and expiry.
	c.mu.RLock()
	entry, ok := c.entries[key]
	c.mu.RUnlock()

	if !ok {
		return "", false
	}

	if time.Now().After(entry.ExpiresAt) {
		// Entry is stale — evict and trigger re-scrape.
		c.mu.Lock()
		// Re-check under write lock to avoid a race where another goroutine
		// already evicted or refreshed the entry.
		if e, still := c.entries[key]; still && time.Now().After(e.ExpiresAt) {
			delete(c.entries, key)
		}
		c.mu.Unlock()
		go c.triggerRescrape(key)
		return "", false
	}

	return entry.Value, true
}

// Delete removes an entry from the cache.
func (c *Cache) Delete(key string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	delete(c.entries, key)
}

// triggerRescrape fires an async HTTP POST to webIntelURL/rescrape with {"key": key}.
// Errors are logged but never panic.
func (c *Cache) triggerRescrape(key string) {
	defer func() {
		if r := recover(); r != nil {
			log.Printf("triggerRescrape panic recovered for key %q: %v", key, r)
		}
	}()

	body, err := json.Marshal(map[string]string{"key": key})
	if err != nil {
		log.Printf("triggerRescrape: failed to marshal payload for key %q: %v", key, err)
		return
	}

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Post(c.webIntelURL+"/rescrape", "application/json", bytes.NewReader(body))
	if err != nil {
		log.Printf("triggerRescrape: HTTP POST failed for key %q: %v", key, err)
		return
	}
	defer resp.Body.Close()
	log.Printf("triggerRescrape: re-scrape triggered for key %q, status %s", key, resp.Status)
}

File: services/kg_cache/cache.py
import logging
import threading
import time
from typing import Optional

import httpx

logger = logging.getLogger(__name__)

TTL_SECONDS = 15 * 60  # 15 minutes (Req 11.4)


class _CacheEntry:
    __slots__ = ("value", "expires_at")

    def __init__(self, value: str, expires_at: float):
        self.value = value
        self.expires_at = expires_at


class TTLCache:
    def __init__(self, web_intel_url: str):
        self._lock = threading.Lock()
        self._entries: dict[str, _CacheEntry] = {}
        self._web_intel_url = web_intel_url

    def set(self, key: str, value: str) -> None:
        with self._lock:
            self._entries[key] = _CacheEntry(value=value, expires_at=time.monotonic() + TTL_SECONDS)

    def get(self, key: str) -> Optional[str]:
        with self._lock:
            entry = self._entries.get(key)
            if entry is None:
                return None
            if time.monotonic() > entry.expires_at:
                del self._entries[key]
                threading.Thread(
                    target=self._trigger_rescrape, args=(key,), daemon=True
                ).start()
                return None
            return entry.value

    def delete(self, key: str) -> None:
        with self._lock:
            self._entries.pop(key, None)

    def _trigger_rescrape(self, key: str) -> None:
        try:
            with httpx.Client(timeout=10.0) as client:
                resp = client.post(
                    f"{self._web_intel_url}/rescrape",
                    json={"key": key},
                )
            logger.info("Rescrape triggered key=%s status=%d", key, resp.status_code)
        except Exception as e:
            logger.error("Rescrape failed key=%s error=%s", key, e)
